restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java

```diff
@@ -183,34 +308,52 @@ public RawDocument toDoc() {
         throw new IllegalStateException("Incomplete document.");
       }
 
-      boolean hasNext = next != null || lastResultSet.getPagingState() != null;
-      return new RawDocument(id, docKey, lastResultSet, hasNext, rows);
+      List<Row> docRows = this.rows.stream().map(DocProperty::row).collect(Collectors.toList());
+
+      CombinedPagingState combinedPagingState = new CombinedPagingState(pagingState);
+
+      return new RawDocument(id, docKey, combinedPagingState, docRows);
     }
 
-    private Accumulator end() {
+    private Accumulator complete(PagingStateTracker tracker, Accumulator next) {
+      rows.forEach(tracker::track);
+      List<PagingStateSupplier> currentPagingState = tracker.slice();
+
+      return new Accumulator(this, currentPagingState, next);
+    }
+
+    private Accumulator end(PagingStateTracker tracker) {
       if (next != null) {
         if (!complete) {
           throw new IllegalStateException("Ending an incomplete document");
         }
 
-        return next.end();
+        return next.end(tracker);
       }
 
       if (complete) {
         throw new IllegalStateException("Already complete");
       }
 
-      return new Accumulator(id, docKey, lastResultSet, rows, null);
+      return complete(tracker, null);
     }
 
     private void append(Accumulator other) {
-      rows.addAll(other.rows);
-      lastResultSet = other.lastResultSet;
+      if (rowComparator == null || lastRow == null) {
+        throw new IllegalStateException("Cannot compare");
+      }
+
+      DocProperty otherRow = other.lastRow;
+
+      if (rowComparator.compare(this.lastRow, otherRow) != 0) {
+        rows.add(otherRow);
+        this.lastRow = otherRow;
```

Comment:
so is this correct, you'll always update the last row, no matter if the comparator returns `1` or `-1`? How's that?

Reply:
We only care about equality at this point.

I hope it's a bit clearer in the refactored code. Please see the latest commits.